flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/batch/BatchExecAggRuleBase.scala

```diff
@@ -0,0 +1,176 @@
+      ).toList
+
+    typeFactory.createStructType(localAggFieldTypes, localAggFieldNames)
+  }
+
+  protected def isTwoPhaseAggWorkable(
+      aggFunctions: Array[UserDefinedFunction],
+      tableConfig: TableConfig): Boolean = {
+    getAggEnforceStrategy(tableConfig) match {
+      case AggPhaseEnforcer.ONE_PHASE => false
+      case _ => doAllSupportMerge(aggFunctions)
+    }
+  }
+
+  protected def isOnePhaseAggWorkable(
+      agg: Aggregate,
+      aggFunctions: Array[UserDefinedFunction],
+      tableConfig: TableConfig): Boolean = {
+    getAggEnforceStrategy(tableConfig) match {
+      case AggPhaseEnforcer.ONE_PHASE => true
+      case AggPhaseEnforcer.TWO_PHASE => !doAllSupportMerge(aggFunctions)
+      case AggPhaseEnforcer.NONE =>
+        if (!doAllSupportMerge(aggFunctions)) {
+          true
+        } else {
+          // if ndv of group key in aggregate is Unknown and all aggFunctions are splittable,
+          // use two-phase agg.
+          // else whether choose one-phase agg or two-phase agg depends on CBO.
+          val mq = agg.getCluster.getMetadataQuery
+          mq.getDistinctRowCount(agg.getInput, agg.getGroupSet, null) != null
+        }
+    }
+  }
+
+  protected def doAllSupportMerge(aggFunctions: Array[UserDefinedFunction]): Boolean = {
+    val supportLocalAgg = aggFunctions.forall {
+      case _: DeclarativeAggregateFunction => true
+      case a => ifMethodExistInFunction("merge", a)
+    }
+    //it means grouping without aggregate functions
+    aggFunctions.isEmpty || supportLocalAgg
+  }
+
+  protected def isEnforceOnePhaseAgg(tableConfig: TableConfig): Boolean = {
+    getAggEnforceStrategy(tableConfig) == AggPhaseEnforcer.ONE_PHASE
+  }
+
+  protected def isEnforceTwoPhaseAgg(tableConfig: TableConfig): Boolean = {
+    getAggEnforceStrategy(tableConfig) == AggPhaseEnforcer.TWO_PHASE
+  }
+
+  protected def getAggEnforceStrategy(tableConfig: TableConfig): AggPhaseEnforcer.Value = {
+    val aggPrefConfig = tableConfig.getConf.getString(
+      PlannerConfigOptions.SQL_OPTIMIZER_AGG_PHASE_ENFORCER)
+    AggPhaseEnforcer.values.find(_.toString.equalsIgnoreCase(aggPrefConfig))
+      .getOrElse(throw new IllegalArgumentException(
+        "Agg phase enforcer can only set to be: NONE, ONE_PHASE, TWO_PHASE!"))
+  }
+
+  protected def isAggBufferFixedLength(agg: Aggregate): Boolean = {
+    val (_, aggCallsWithoutAuxGroupCalls) = FlinkRelOptUtil.checkAndSplitAggCalls(agg)
+    val (_, aggBufferTypes, _) = AggregateUtil.transformToBatchAggregateFunctions(
+      aggCallsWithoutAuxGroupCalls, agg.getInput.getRowType)
+
+    isAggBufferFixedLength(aggBufferTypes.map(_.map(TypeConverters.createInternalTypeFromTypeInfo)))
+  }
+
+  protected def isAggBufferFixedLength(aggBufferTypes: Array[Array[InternalType]]): Boolean = {
+    val aggBuffAttributesTypes = aggBufferTypes.flatten
+    val isAggBufferFixedLength = aggBuffAttributesTypes.forall(BinaryRow.isMutable)
```

Comment:
add a TODO to check `Decimal` can be applied to hash agg, like `sum(decimal)`

Reply:
I have added test case to verify `Decimal ` type in `AggregateTestBase#testSum`